### PR TITLE
Enable nullability in CodeDomSerializerBase

### PR DESCRIPTION
**Pull request by halgab**
This PR is pretty big, apologies for that. It enables nullability in CodeDomSerializerBase to get one foot in the door to annotate more serializers down the road,
One issue I had doing this is the lack of annoations in System.CodeDom (see https://github.com/dotnet/runtime/issues/41720, https://github.com/dotnet/runtime/issues/78036)

Contributes to #8342

## Proposed changes

- Enable nullability in CodeDomSerializerBase
- Refactor CodeDomSerializerBase and fix an apparent bug
- Introduce new helper functions to simplify more parts of the code

Each change is its own commit.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9099)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 1]
```diff
@@ -508,62 +484,42 @@ protected void DeserializeStatement(IDesignerSerializationManager manager, CodeS
...
-                            CodeExpressionStatement ces = statement as CodeExpressionStatement;
-                            if (ces is not null)
-                            {
-                                DeserializeExpression(manager, null, ces.Expression);
-                            }
-                            else
-                            {
-                                if (statement is CodeMethodReturnStatement cmrs)
-                                {
-                                    DeserializeExpression(manager, null, ces.Expression);
```
Comment (halgab) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs:538]: This looked a lot like a bug

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 2]
```diff
@@ -1243,11 +1167,11 @@ protected object DeserializeExpression(IDesignerSerializationManager manager, st
                             //
                             for (int i = 0; i < indexes.Length; i++)
                             {
-                                IConvertible index = DeserializeExpression(manager, name, arrayIndexerEx.Indices[i]) as IConvertible;
+                                object? index = DeserializeExpression(manager, name, arrayIndexerEx.Indices[i]);
```
Comment (halgab) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs:1170]: This should fix the trace in the non `IConvertible` case

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 3]
```diff
@@ -507,62 +482,42 @@ protected void DeserializeStatement(IDesignerSerializationManager manager, CodeS
...
-                        CodeExpressionStatement ces = statement as CodeExpressionStatement;
-                        if (ces is not null)
-                        {
-                            DeserializeExpression(manager, null, ces.Expression);
-                        }
-                        else
-                        {
-                            if (statement is CodeMethodReturnStatement cmrs)
-                            {
-                                DeserializeExpression(manager, null, ces.Expression);
```
Comment (halgab) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs:537]: This looked a lot like a bug

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 4]
```diff
@@ -1242,11 +1165,11 @@ protected object DeserializeExpression(IDesignerSerializationManager manager, st
                         //
                         for (int i = 0; i < indexes.Length; i++)
                         {
-                            IConvertible index = DeserializeExpression(manager, name, arrayIndexerEx.Indices[i]) as IConvertible;
+                            object? index = DeserializeExpression(manager, name, arrayIndexerEx.Indices[i]);
```
Comment (halgab) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs:1168]: This should fix the trace in the non IConvertible case

**Comment by elachlan** [comment]
Thanks for working on this!

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 5]
```diff
@@ -1488,73 +1411,48 @@ private void DeserializeAttachEventStatement(IDesignerSerializationManager manag
...
-    }
-
     private static object ExecuteBinaryOperator(IConvertible left, IConvertible right, CodeBinaryOperatorType op)
     {
         TypeCode leftType = left.GetTypeCode();
         TypeCode rightType = right.GetTypeCode();
 
         // The compatible types are listed in order from lowest bitness to highest.  We must operate on the highest bitness to keep fidelity.
-        TypeCode[] compatibleTypes = new TypeCode[]
+        ReadOnlySpan<TypeCode> compatibleTypes = new TypeCode[]
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs:1455]: Why change this to ReadOnlySpan?

File: src/System.Windows.Forms.Primitives/src/System/ComponentModel/Design/Serialization/DesignerSerializationManagerHelper.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.ComponentModel.Design.Serialization
+{
+    internal static class DesignerSerializationManagerHelper
+    {
+        public static bool TryGetContext<T>(this IDesignerSerializationManager manager,
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms.Primitives/src/System/ComponentModel/Design/Serialization/DesignerSerializationManagerHelper.cs:9]: nit: when we break parameters they should all start indented on their own line

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
In general, looks really good! I think there is a bit more opportunity to use pattern matching. See my comments.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 7]
```diff
@@ -56,63 +55,58 @@ internal static string GetTypeNameFromCodeTypeReference(IDesignerSerializationMa
...
 
-        //now create each sub-argument part.
-        foreach (CodeTypeReference childref in typeref.TypeArguments)
-        {
-            if (!first)
+            //now create each sub-argument part.
+            foreach (CodeTypeReference childref in typeref.TypeArguments)
             {
-                typename.Append(',');
+                typeName.Append('[');
```
[the remark this answers is not in the document]
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs:92]: Done. In the end there weren't many opportunities to do so, as line 93 isn't a direct `Append` call

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 8]
```diff
@@ -194,7 +180,7 @@ protected static Type GetReflectionTypeHelper(IDesignerSerializationManager mana
         Type type = instance.GetType();
         if (type.IsValueType)
         {
-            TypeDescriptionProvider targetProvider = GetTargetFrameworkProvider(manager, instance);
+            TypeDescriptionProvider? targetProvider = GetTargetFrameworkProvider(manager, instance);
```
[the remark this answers is not in the document]
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs]: I took the opportunity to move this logic to a helper to reduce code duplication

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs [hunk 9]
```diff
@@ -222,13 +208,13 @@ protected static PropertyDescriptorCollection GetPropertiesHelper(IDesignerSeria
...
             if (targetProvider is not null)
             {
                 // target framework provider is null at runtime
                 if (targetProvider.IsSupportedType(instance.GetType()))
                 {
-                    ICustomTypeDescriptor targetAwareDescriptor = targetProvider.GetTypeDescriptor(instance);
+                    ICustomTypeDescriptor? targetAwareDescriptor = targetProvider.GetTypeDescriptor(instance);
```
[the remark this answers is not in the document]
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomSerializerBase.cs]: done

File: src/System.Windows.Forms.Primitives/src/System/ComponentModel/Design/Serialization/DesignerSerializationManagerHelper.cs (hunk 6 above)
Comment (halgab) [reply src/System.Windows.Forms.Primitives/src/System/ComponentModel/Design/Serialization/DesignerSerializationManagerHelper.cs]: done

**Review by JeremyKuhne — APPROVED** [review]
Looks good, thanks for this!

**Comment by steveharter** [comment]
What is left to fully address https://github.com/dotnet/winforms/issues/8342?

**Comment by halgab** [comment]
There are quite a lot of files that still need annotating. I count 192 as of current main branch.

**Pull request merged**